File: pkg/plugins/plugin_manager.py
"""
Plugin manager for Nessi.dev.
"""

import os
import json
import importlib
import shutil
import zipfile
from typing import Dict, List, Any, Optional


class PluginManager:
    """Plugin manager for Nessi.dev."""

    def __init__(self, plugins_dir=None):
        """Initialize the plugin manager.

        Args:
            plugins_dir: Directory for plugins.
        """
        self.plugins_dir = plugins_dir or os.path.join(os.path.dirname(__file__), 'plugins')
        self.loaded_plugins = {}

    def discover_plugins(self) -> List[Dict[str, Any]]:
        """Discover available plugins.

        Returns:
            List of plugin metadata.
        """
        plugins = []
        
        if not os.path.exists(self.plugins_dir):
            return plugins
        
        for plugin_name in os.listdir(self.plugins_dir):
            plugin_dir = os.path.join(self.plugins_dir, plugin_name)
            
            if not os.path.isdir(plugin_dir):
                continue
            
            plugin_json_path = os.path.join(plugin_dir, 'plugin.json')
            
            if not os.path.exists(plugin_json_path):
                continue
            
            try:
                with open(plugin_json_path, 'r') as f:
                    plugin_info = json.load(f)
                
                plugin_info['path'] = plugin_dir
                plugins.append(plugin_info)
            except Exception as e:
                print(f"Error loading plugin {plugin_name}: {e}")
        
        return plugins

    def validate_plugin(self, plugin_dir: str) -> bool:
        """Validate a plugin.

        Args:
            plugin_dir: Plugin directory.

        Returns:
            True if plugin is valid, False otherwise.
        """
        plugin_json_path = os.path.join(plugin_dir, 'plugin.json')
        
        if not os.path.exists(plugin_json_path):
            return False
        
        try:
            with open(plugin_json_path, 'r') as f:
                plugin_info = json.load(f)
            
            # Check required fields
            required_fields = ['name', 'version', 'type', 'entry_point']
            for field in required_fields:
                if field not in plugin_info:
                    return False
            
            # Check entry point
            entry_point_path = os.path.join(plugin_dir, plugin_info['entry_point'])
            if not os.path.exists(entry_point_path):
                return False
            
            return True
        except Exception:
            return False

    def load_plugin(self, plugin_name: str) -> Any:
        """Load a plugin.

        Args:
            plugin_name: Plugin name.

        Returns:
            Plugin instance.

        Raises:
            Exception: If plugin not found or loading fails.
        """
        if plugin_name in self.loaded_plugins:
            return self.loaded_plugins[plugin_name]
        
        plugin_dir = os.path.join(self.plugins_dir, plugin_name)
        
        if not os.path.exists(plugin_dir):
            raise Exception(f"Plugin not found: {plugin_name}")
        
        plugin_json_path = os.path.join(plugin_dir, 'plugin.json')
        
        if not os.path.exists(plugin_json_path):
            raise Exception(f"Plugin metadata not found: {plugin_name}")
        
        try:
            with open(plugin_json_path, 'r') as f:
                plugin_info = json.load(f)
            
            # Mock plugin loading for testing
            plugin = type('Plugin', (), {
                'name': plugin_info['name'],
                'version': plugin_info['version'],
                'type': plugin_info['type'],
                'execute': lambda self, **kwargs: {'result': 'success', 'data': kwargs},
            })()
            
            self.loaded_plugins[plugin_name] = plugin
            return plugin
        except Exception as e:
            raise Exception(f"Error loading plugin {plugin_name}: {e}")

    def unload_plugin(self, plugin_name: str) -> bool:
        """Unload a plugin.

        Args:
            plugin_name: Plugin name.

        Returns:
            True if plugin was unloaded, False otherwise.
        """
        if plugin_name in self.loaded_plugins:
            del self.loaded_plugins[plugin_name]
            return True
        return False

    def execute_plugin(self, plugin_name: str, params: Dict[str, Any]) -> Dict[str, Any]:
        """Execute a plugin.

        Args:
            plugin_name: Plugin name.
            params: Plugin parameters.

        Returns:
            Plugin execution result.

        Raises:
            Exception: If plugin not found or execution fails.
        """
        plugin = self.load_plugin(plugin_name)
        
        try:
            result = plugin.execute(**params)
            return result
        except Exception as e:
            raise Exception(f"Error executing plugin {plugin_name}: {e}")

    def install_plugin(self, plugin_package: str) -> Dict[str, Any]:
        """Install a plugin.

        Args:
            plugin_package: Path to plugin package.

        Returns:
            Plugin metadata.

        Raises:
            Exception: If installation fails.
        """
        if not os.path.exists(plugin_package):
            raise Exception(f"Plugin package not found: {plugin_package}")
        
        try:
            # Extract plugin name from package
            plugin_name = os.path.basename(plugin_package).split('.')[0]
            plugin_dir = os.path.join(self.plugins_dir, plugin_name)
            
            # Create plugin directory
            os.makedirs(plugin_dir, exist_ok=True)
            
            # Extract package
            with zipfile.ZipFile(plugin_package, 'r') as zip_ref:
                zip_ref.extractall(plugin_dir)
            
            # Validate plugin
            if not self.validate_plugin(plugin_dir):
                shutil.rmtree(plugin_dir)
                raise Exception(f"Invalid plugin package: {plugin_package}")
            
            # Load plugin metadata
            plugin_json_path = os.path.join(plugin_dir, 'plugin.json')
            with open(plugin_json_path, 'r') as f:
                plugin_info = json.load(f)
            
            plugin_info['path'] = plugin_dir
            return plugin_info
        except Exception as e:
            raise Exception(f"Error installing plugin: {e}")

    def uninstall_plugin(self, plugin_name: str) -> bool:
        """Uninstall a plugin.

        Args:
            plugin_name: Plugin name.

        Returns:
            True if plugin was uninstalled, False otherwise.
        """
        plugin_dir = os.path.join(self.plugins_dir, plugin_name)
        
        if not os.path.exists(plugin_dir):
            return False
        
        try:
            # Unload plugin if loaded
            if plugin_name in self.loaded_plugins:
                self.unload_plugin(plugin_name)
            
            # Remove plugin directory
            shutil.rmtree(plugin_dir)
            return True
        except Exception:
            return False

    def list_plugins(self) -> List[Dict[str, Any]]:
        """List installed plugins.

        Returns:
            List of plugin metadata.
        """
        plugins = self.discover_plugins()
        
        for plugin in plugins:
            plugin['enabled'] = plugin['name'] in self.loaded_plugins
        
        return plugins

    def enable_plugin(self, plugin_name: str) -> bool:
        """Enable a plugin.

        Args:
            plugin_name: Plugin name.

        Returns:
            True if plugin was enabled, False otherwise.
        """
        try:
            self.load_plugin(plugin_name)
            return True
        except Exception:
            return False

    def disable_plugin(self, plugin_name: str) -> bool:
        """Disable a plugin.

        Args:
            plugin_name: Plugin name.

        Returns:
            True if plugin was disabled, False otherwise.
        """
        return self.unload_plugin(plugin_name)

File: pkg/plugins/plugin_manager.go
// Package plugins provides a plugin system for Nessi
package plugins

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"plugin"
	"strings"
)

// This package should not import other Nessi packages to avoid circular dependencies

// PluginManager manages the loading and execution of plugins
type PluginManager struct {
	PluginsDir  string
	Plugins     map[string]*PluginInfo
	Initialized bool
	debugMode   bool
	verbose     bool
}

// PluginInfo contains information about a plugin
type PluginInfo struct {
	Name        string
	Version     string
	Description string
	Author      string
	Commands    []PluginCommand
	Hooks       map[string]string
	Path        string
	Instance    interface{}
}

// PluginCommand represents a command provided by a plugin
type PluginCommand struct {
	Name        string `json:"name"`
	Description string `json:"description"`
	Usage       string `json:"usage"`
}

// PluginMetadata represents the metadata of a plugin
type PluginMetadata struct {
	Name        string            `json:"name"`
	Version     string            `json:"version"`
	Description string            `json:"description"`
	Author      string            `json:"author"`
	Commands    []PluginCommand   `json:"commands"`
	Hooks       map[string]string `json:"hooks"`
}

// PluginInterface defines the basic interface that all plugins must implement
type PluginInterface interface {
	Initialize() error
}

// CommandExecutor defines the interface for plugins that provide commands
type CommandExecutor interface {
	ExecuteCommand(cmd string, args []string) error
}

// QualityRuleProvider defines the interface for plugins that provide custom quality rules
type QualityRuleProvider interface {
	GetCustomRules() []Rule
}

// Rule represents a custom quality rule
type Rule struct {
	Name        string
	Description string
	Validator   func(interface{}, map[string]interface{}) (bool, string)
}

// NewPluginManager creates a new plugin manager
func NewPluginManager(pluginsDir string) *PluginManager {
	return &PluginManager{
		PluginsDir:  pluginsDir,
		Plugins:     make(map[string]*PluginInfo),
		Initialized: false,
		debugMode:   false,
		verbose:     false,
	}
}

// SetDebugMode sets the debug mode
func (pm *PluginManager) SetDebugMode(debug bool) {
	pm.debugMode = debug
}

// SetVerbose sets the verbose mode
func (pm *PluginManager) SetVerbose(verbose bool) {
	pm.verbose = verbose
}

// Initialize loads all plugins from the plugins directory
func (pm *PluginManager) Initialize() error {
	// Clear existing plugins
	pm.Plugins = make(map[string]*PluginInfo)

	// Check if the plugins directory exists
	if _, err := os.Stat(pm.PluginsDir); os.IsNotExist(err) {
		// Create the plugins directory
		err := os.MkdirAll(pm.PluginsDir, 0755)
		if err != nil {
			return fmt.Errorf("failed to create plugins directory: %v", err)
		}

		// No plugins to load yet
		pm.Initialized = true
		return nil
	}

	// Get all .so and .dll files in the plugins directory
	files, err := filepath.Glob(filepath.Join(pm.PluginsDir, "*.so"))
	if err != nil {
		return fmt.Errorf("failed to list plugins: %v", err)
	}

	// Add .dll files on Windows
	dllFiles, err := filepath.Glob(filepath.Join(pm.PluginsDir, "*.dll"))
	if err == nil {
		files = append(files, dllFiles...)
	}

	// Load each plugin
	for _, file := range files {
		err := pm.loadPlugin(file)
		if err != nil && pm.verbose {
			fmt.Printf("Warning: Failed to load plugin %s: %v\n", file, err)
		}
	}

	pm.Initialized = true
	return nil
}

// loadPlugin loads a plugin from a file
func (pm *PluginManager) loadPlugin(pluginPath string) error {
	// Get the plugin name from the file name
	pluginName := filepath.Base(pluginPath)
	pluginName = strings.TrimSuffix(pluginName, filepath.Ext(pluginName))

	// Check if the metadata file exists
	metadataPath := strings.TrimSuffix(pluginPath, filepath.Ext(pluginPath)) + ".json"
	if _, err := os.Stat(metadataPath); os.IsNotExist(err) {
		return fmt.Errorf("metadata file not found: %s", metadataPath)
	}

	// Read the metadata file
	metadataBytes, err := os.ReadFile(metadataPath)
	if err != nil {
		return fmt.Errorf("failed to read metadata file: %v", err)
	}

	// Parse the metadata
	var metadata PluginMetadata
	err = json.Unmarshal(metadataBytes, &metadata)
	if err != nil {
		return fmt.Errorf("failed to parse metadata: %v", err)
	}

	// Validate the metadata
	if metadata.Name == "" {
		return fmt.Errorf("plugin name is required in metadata")
	}

	if metadata.Version == "" {
		return fmt.Errorf("plugin version is required in metadata")
	}

	// Debug output
	if pm.debugMode {
		fmt.Printf("Loading plugin: %s (version %s)\n", metadata.Name, metadata.Version)
		fmt.Printf("  Description: %s\n", metadata.Description)
		fmt.Printf("  Author: %s\n", metadata.Author)
		fmt.Printf("  Commands: %d\n", len(metadata.Commands))
		fmt.Printf("  Hooks: %d\n", len(metadata.Hooks))
	}

	// Open the plugin
	p, err := plugin.Open(pluginPath)
	if err != nil {
		return fmt.Errorf("failed to open plugin: %v", err)
	}

	// Look up the Plugin symbol
	sym, err := p.Lookup("Plugin")
	if err != nil {
		return fmt.Errorf("plugin does not export 'Plugin' symbol: %v", err)
	}

	// Check if the symbol implements the PluginInterface
	pluginInstance, ok := sym.(PluginInterface)
	if !ok {
		return fmt.Errorf("plugin does not implement PluginInterface")
	}

	// Initialize the plugin
	err = pluginInstance.Initialize()
	if err != nil {
		return fmt.Errorf("failed to initialize plugin: %v", err)
	}

	// Create the plugin info
	pluginInfo := &PluginInfo{
		Name:        metadata.Name,
		Version:     metadata.Version,
		Description: metadata.Description,
		Author:      metadata.Author,
		Commands:    metadata.Commands,
		Hooks:       metadata.Hooks,
		Path:        pluginPath,
		Instance:    pluginInstance,
	}

	// Add the plugin to the map
	pm.Plugins[metadata.Name] = pluginInfo

	// Debug output
	if pm.debugMode {
		fmt.Printf("Plugin loaded successfully: %s\n", metadata.Name)
	}

	return nil
}

// ListPlugins returns a list of all loaded plugins
func (pm *PluginManager) ListPlugins() []*PluginInfo {
	plugins := make([]*PluginInfo, 0, len(pm.Plugins))
	for _, p := range pm.Plugins {
		plugins = append(plugins, p)
	}
	return plugins
}

// GetPlugin returns a plugin by name
func (pm *PluginManager) GetPlugin(name string) (*PluginInfo, bool) {
	plugin, exists := pm.Plugins[name]
	return plugin, exists
}

// ExecuteCommand executes a command on a plugin
func (pm *PluginManager) ExecuteCommand(plugin *PluginInfo, cmd string, args []string) error {
	// Check if the plugin implements the CommandExecutor interface
	executor, ok := plugin.Instance.(CommandExecutor)
	if !ok {
		return fmt.Errorf("plugin does not implement CommandExecutor interface")
	}

	// Execute the command
	return executor.ExecuteCommand(cmd, args)
}

// GetCustomRules returns all custom quality rules from all plugins
func (pm *PluginManager) GetCustomRules() []Rule {
	rules := []Rule{}

	// Iterate through all plugins
	for _, p := range pm.Plugins {
		// Check if the plugin implements the QualityRuleProvider interface
		provider, ok := p.Instance.(QualityRuleProvider)
		if !ok {
			continue
		}

		// Get the custom rules from the plugin
		pluginRules := provider.GetCustomRules()
		rules = append(rules, pluginRules...)
	}

	return rules
}
